### PR TITLE
Add support for #derive[FromStr] for enums

### LINKED ISSUE #59: Implement FromStr for Enum
I have an enum with a long list like this:

```rust
enum ThingType {
    A,
    B,
    ...
    Z,
}
```

I want to be able to do

```rust
ThingType::from_str("A")
```
and get `ThingType::A`.

### PR DESCRIPTION
**Pull request by aj-bagwell**
This allows deriving FromStr for enums with no fields in the obvious way

fixes #59

### CODE REVIEW

File: src/from_str.rs [hunk 1]
```diff
@@ -41,7 +52,81 @@ pub fn expand(input: &DeriveInput, trait_name: &'static str) -> Result<TokenStre
...
+        }
+
+        impl std::error::Error for #err_name {}
+
+        impl #trait_path for #input_type
+        {
+            type Err = #err_name;
+            #[inline]
+            fn from_str(src: &str) -> ::core::result::Result<Self, Self::Err> {
+                Ok(match src.to_lowercase().as_str() {
```
Comment (aj-bagwell) [inline src/from_str.rs:123]: I considered adding `.replace(&[' ', '_', '-'], "")` here so that it would mach against kebab-case and SNAKE_CASE too.

File: src/from_str.rs [hunk 2]
```diff
@@ -41,7 +52,81 @@ pub fn expand(input: &DeriveInput, trait_name: &'static str) -> Result<TokenStre
...
+            }
+        }
+    }
+
+    let trait_path = state.trait_path;
+
+    quote! {
+
+        #[derive(Debug, Clone, PartialEq, Eq)]
+        #visibility struct #err_name;
```
Comment (aj-bagwell) [inline src/from_str.rs:108]: I'm not sure I like creating a type out of nothing here, but I did not like the other options better:

- could use `&'static str` but that does not impelement `std::error::Error` which make it annoying for combining with other errors
- could have a `derive_more::ParseError` but that makes the dependencies complicates as you cant have code and procmarcos in the same dependency

**Comment by aj-bagwell** [comment]
Forced push changes that hopefull fix the failing tests

File: Cargo.toml [hunk 3]
```diff
@@ -51,7 +51,7 @@ deref_mut = []
 display = ["syn/extra-traits"]
 error = ["syn/extra-traits"]
 from = ["syn/extra-traits"]
-from_str = []
+from_str = ["convert_case"]
```
Comment (JelteF) [inline Cargo.toml:54]: This should fix the current compilation issue in the feature tests.
```suggestion
from_str = ["convert_case", "syn/extra-traits"]
```

Comment (aj-bagwell) [reply Cargo.toml:54]: Thanks, I just furoced pushed a change that bumps `syn` to 1.0.81 (which is the version in the lock file) and changed the code to use .is_empty() instead which means it does not need the extra traits

Comment (aj-bagwell) [reply Cargo.toml:54]: Bah looks like I also need to ditch the `x == from_str(x.to_string())` test as I can't `#derive[Display]` if the display feature is not set.

**Comment by aj-bagwell** [comment]
The test are all now passing and I'm happy with my commits as they stand, @JelteF let me know if there are any more changes that you want me to make.

File: src/from_str.rs [hunk 4]
```diff
@@ -41,7 +51,81 @@ pub fn expand(input: &DeriveInput, trait_name: &'static str) -> Result<TokenStre
...
+            cases.push(quote! {
+                #canonical => #input_type::#variant,
+            })
+        } else {
+            for variant in variants {
+                let variant_str = variant.to_string();
+                cases.push(quote! {
+                    #canonical if(src == #variant_str) => #input_type::#variant,
+                })
+            }
```
Comment (JelteF) [inline src/from_str.rs:98]: this edge case is missing a test

**Review by JelteF — DISMISSED** [review]
I think this is good except for a missing test for the smart case sensitive behaviour.

**Comment by aj-bagwell** [comment]
I have added `enum_test_case_sensitive` to check if falls back to case sentive matching when it matters and rebased onto master to bring it up to date

**Comment by JelteF** [comment]
Thanks a lot for your work on this!
